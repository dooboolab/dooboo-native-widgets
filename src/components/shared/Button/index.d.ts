import * as React from 'react';
import {
  ViewStyle,
  ImageSourcePropType,
  TextStyle,
  ImageStyle,
} from 'react-native';

interface Props {
  testID?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  onClick?: () => void;
  style?: ViewStyle;
  disabledStyle?: ViewStyle;
  textStyle?: TextStyle;
  imgLeftSrc?: ImageSourcePropType;
  imgLeftStyle?: ImageStyle;
  indicatorColor?: string;
  activeOpacity?: number;
  text?: string;
}

const Button: React.SFC<Props> = (props) => {

};

export default Button;
