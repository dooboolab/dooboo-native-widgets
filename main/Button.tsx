import {
  ActivityIndicator,
  LayoutRectangle,
  StyleProp,
  StyleSheet,
  Text,
  TextProps,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle,
} from 'react-native';
import React, {useRef, useState} from 'react';

import {useHover} from 'react-native-web-hooks';

const defaultStyles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    paddingHorizontal: 12,
    paddingVertical: 8,
    borderRadius: 6,

    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    fontSize: 14,
    color: '#069ccd',
  },
  disabledButton: {
    backgroundColor: '#cccccc',
    borderColor: 'rgb(200, 200, 200)',
  },
  disabledText: {
    color: '#969696',
  },
  hovered: {
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.24,
    shadowRadius: 16,
    elevation: 10,
  },
});

type StylesType = Partial<StyleSheet.NamedStyles<typeof defaultStyles>>;

interface Props {
  testID?: string;
  indicatorColor?: string;
  loading?: boolean;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
  styles?: StylesType;
  leftElement?: React.ReactElement;
  rightElement?: React.ReactElement;
  activeOpacity?: TouchableOpacityProps['activeOpacity'];
  text?: string;
  onPress?: TouchableOpacityProps['onPress'];
  touchableOpacityProps?: Partial<TouchableOpacityProps>;
  textProps?: Partial<TextProps>;
}

function Button({
  testID,
  disabled,
  loading,
  style,
  styles,
  indicatorColor = '#ffffff',
  leftElement,
  rightElement,
  activeOpacity = 0.7,
  text,
  onPress,
  touchableOpacityProps,
  textProps,
}: Props): React.ReactElement {
  const ref = useRef<TouchableOpacity>(null);
  const hovered = useHover(ref);
  const [layout, setLayout] = useState<LayoutRectangle>();

  return (
    <TouchableOpacity
      testID={testID}
      ref={ref}
      activeOpacity={activeOpacity}
      onPress={onPress}
      delayPressIn={50}
      disabled={disabled}
      style={style}
      {...touchableOpacityProps}>
      {loading ? (
        <View
          style={[
            defaultStyles.container,
            styles?.container,
            {
              width: layout?.width,
              height: layout?.height,
            },
            hovered && !disabled && [defaultStyles.hovered, styles?.hovered],
            disabled && [defaultStyles.disabledButton, styles?.disabledButton],
          ]}>
          <ActivityIndicator size="small" color={indicatorColor} />
        </View>
      ) : (
        <View
          style={[
            defaultStyles.container,
            styles?.container,
            hovered && !disabled && [defaultStyles.hovered, styles?.hovered],
            disabled && [defaultStyles.disabledButton, styles?.disabledButton],
          ]}
          onLayout={(e) => setLayout(e.nativeEvent.layout)}>
          {leftElement}
          <Text
            style={[
              defaultStyles.text,
              styles?.text,
              disabled && [defaultStyles.disabledText, styles?.disabledText],
            ]}
            {...textProps}>
            {text}
          </Text>
          {rightElement}
        </View>
      )}
    </TouchableOpacity>
  );
}

export {Button};
