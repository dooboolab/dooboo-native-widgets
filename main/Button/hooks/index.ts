import { MutableRefObject } from 'react';
import customHooks from './customHooks';

export const useHover = customHooks({
  events: ['mouseenter', 'mouseleave'],
}) as <T>(ref: MutableRefObject<T>) => boolean;
