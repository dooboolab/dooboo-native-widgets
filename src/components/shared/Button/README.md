# Button
> Simple [Button] component that can be used inside product. Has basic features like `loading` state, `disabled` state and also has ability to put `img` to left-hand which is used very often.

![image](https://user-images.githubusercontent.com/27461460/62291727-9be84100-b49f-11e9-8ce5-ceaa1dc3153e.png)

## Props
|                      | necessary | types                  | default                      |
|----------------------|-----------|------------------------|------------------------------|
| testID               |           | string                 |                              |
| isLoading            |           | boolean                |                              |
| isDisabled           |           | boolean                |                              |
| onClick              |           | func                   |                              |
| style                |           | `StyleProp<ViewStyle>` |                              |
| disabledStyle        |           | `StyleProp<ViewStyle>` |                              |
| textStyle            |           | `StyleProp<TextStyle>` |                              |
| imgLeftSrc           |           | ImageSourcePropType    |                              |
| imgLeftStyle         |           | ImageSourcePropType    |                              |
| indicatorColor       |           | string                 |                              |
| activeOpacity        |           | number                 |                              |
| text                 |           | string                 |                              |

## Getting started

* Import
  ```javascript
  import { Button } from '@dooboo-ui/native';
  ```

* Usage
  ```javascript
  function Page(props: Props) {
    return (
      <Container>
        <Button
          testID='btn'
          isLoading={false}
          text='😀 😎 👍 💯'
          onClick={() => {}}
        />
        <Button
          style={{
            marginVertical: 40,
          }}
          isDisabled={true}
          text='This is disabled!!'
          onClick={() => {}}
        />
        <Button
          testID='btnGoogle'
          imgLeftSrc={IC_GOOGLE}
          isLoading={googleLoading}
          indicatorColor='#023059'
          onClick={() => {
            setGoogleLoading(true);
            const timeout = setTimeout(() => {
              setGoogleLoading(false);
              clearTimeout(timeout);
            }, 2000);
          }}
          text='GOOGLE SIGN IN'
        />
        <Button
          testID='btnFacebook'
          imgLeftSrc={IC_FACEBOOK}
          indicatorColor='#023059'
          isLoading={facebookLoading}
          imgLeftStyle={{
            height: 28,
            width: 16,
          }}
          style={{
            marginTop: 40,
            backgroundColor: '#ccc',
            borderWidth: 0.5,
            borderRadius: 0,
          }}
          onClick={() => {
            setFacebookLoading(true);
            const timeout = setTimeout(() => {
              setFacebookLoading(false);
              clearTimeout(timeout);
            }, 2000);
          }}
          text='FACEBOOK SIGN IN'
        />
      </Container>
    );
  }
```
