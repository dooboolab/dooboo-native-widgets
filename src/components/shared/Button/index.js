import React, { Component } from 'react';
import {
  ActivityIndicator,
  TouchableOpacity,
  View,
  Image,
  Text,
  ViewStyle,
  ViewPropTypes,
  TextStyle,
  ImageStyle,
  StyleSheet,
} from 'react-native';
import PropTypes from 'prop-types';

const styles = StyleSheet.create({
  button: {
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: 320,
    height: 52,
    borderColor: 'blue',
    alignItems: 'center',
    justifyContent: 'center',
  },
  disabledButton: {
    backgroundColor: '#cccccc',
    borderColor: 'rgb(200,200,200)',
  },
  text: {
    fontSize: 14,
    color: '#069ccd',
  },
  textDisabled: {
    color: '#969696',
  },
  image: {
    width: 24,
    height: 24,
    position: 'absolute',
    left: 16,
  },
});

function Button(props) {
  if (props.isDisabled) {
    return (
      <View
        testID={props.testID}
        style={[
          styles.button,
          props.style,
          styles.disabledButton,
          props.disabledStyle,
        ]}
      >
        <Text style={props.textStyle}>{props.text}</Text>
      </View>
    );
  }
  if (props.isLoading) {
    return (
      <View
        testID={props.testID}
        style={[
          styles.button,
          props.style,
        ]}
      >
        <ActivityIndicator size='small' color={props.indicatorColor} />
      </View>
    );
  }
  return (
    <TouchableOpacity
      testID={props.testID}
      activeOpacity={props.activeOpacity}
      onPress={props.onClick}
    >
      <View style={[
        styles.button,
        props.style,
      ]}>
        {
          props.imgLeftSrc
            ? <Image
              style={props.imgLeftStyle}
              source={props.imgLeftSrc}
            />
            : null
        }
        <Text style={props.textStyle}>{props.text}</Text>
      </View>
    </TouchableOpacity>
  );
}

Button.defaultProps = {
  isLoading: false,
  isDisabled: false,
  indicatorColor: 'white',
  activeOpacity: 0.5,
  imgLeftStyle: styles.image,
  textStyle: styles.text,
  disabledStyle: styles.disabledStyle,
};

Button.propTypes = {
  testID: PropTypes.string,
  onClick: PropTypes.func,
  style: ViewPropTypes.style,
  disabledStyle: ViewPropTypes.style,
  textStyle: Text.propTypes.style,
  imgLeftSrc: Image.propTypes.source,
  imgLeftStyle: Image.propTypes.style,
  indicatorColor: PropTypes.string,
  activeOpacity: PropTypes.number,
  text: PropTypes.string,
  isLoading: PropTypes.bool,
  isDisabled: PropTypes.bool,
};

export default Button;
