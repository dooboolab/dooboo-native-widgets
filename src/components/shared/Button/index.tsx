import {
  ActivityIndicator,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';

import React from 'react';

const styles = StyleSheet.create({
  button: {
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: 320,
    height: 52,
    borderColor: 'blue',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  disabledButton: {
    backgroundColor: '#cccccc',
    borderColor: 'rgb(200,200,200)',
  },
  text: {
    fontSize: 14,
    color: '#069ccd',
  },
  textDisabled: {
    color: '#969696',
  },
});

interface Props {
  testID?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  onClick?: () => void;
  style?: StyleProp<ViewStyle>;
  disabledStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  leftComponent?: React.ReactElement;
  indicatorColor?: string;
  activeOpacity?: number;
  text?: string;
}

function Button(props: Props): React.ReactElement {
  if (props.isDisabled) {
    return (
      <View
        testID={props.testID}
        style={[
          styles.button,
          props.style,
          styles.disabledButton,
          props.disabledStyle,
        ]}
      >
        <Text style={props.textStyle}>{props.text}</Text>
      </View>
    );
  }
  if (props.isLoading) {
    return (
      <View testID={props.testID} style={[styles.button, props.style]}>
        <ActivityIndicator size="small" color={props.indicatorColor} />
      </View>
    );
  }
  return (
    <TouchableOpacity
      testID={props.testID}
      activeOpacity={props.activeOpacity}
      onPress={props.onClick}
    >
      <View style={[styles.button, props.style]}>
        {props.leftComponent ? props.leftComponent : null}
        <Text style={props.textStyle}>{props.text}</Text>
      </View>
    </TouchableOpacity>
  );
}

export default Button;
