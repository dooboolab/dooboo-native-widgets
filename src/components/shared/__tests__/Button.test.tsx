import * as React from 'react';

import Button from '../Button';
import { Text } from 'react-native';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const component = (props?: any) => {
  return <Button {...props} />;
};

describe('[Button]', () => {
  let rendered: renderer.ReactTestRenderer;

  it('should render without crashing', () => {
    rendered = renderer.create(component());
    expect(rendered.toJSON()).toMatchSnapshot();
    expect(rendered.toJSON()).toBeTruthy();
  });

  describe('[Button] Interaction', () => {
    let root: renderer.ReactTestInstance;
    let cnt = 1;
    it('simulate onPress', () => {
      rendered = renderer.create(
        component({
          onClick: () => cnt++,
        })
      );
      root = rendered.root;

      root.findByType(Button).props.onClick();
      expect(cnt).toBe(2);
    });

    it('renders disabled', () => {
      rendered = renderer.create(component({ isDisabled: true }));
      root = rendered.root;

      const texts = root.findAllByType(Text);
      expect(texts).toHaveLength(1);
    });
  });
});
