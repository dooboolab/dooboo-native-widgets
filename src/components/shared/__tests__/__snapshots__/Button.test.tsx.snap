// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Button] renders without crashing 1`] = `
<View
  accessible={true}
  clickable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#0000ff",
          "borderRadius": 4,
          "borderWidth": 2,
          "flexDirection": "row",
          "height": 52,
          "justifyContent": "center",
          "width": 320,
        },
        Object {},
      ]
    }
    theme={
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#0000ff",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "left": 0,
            "position": "absolute",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#069ccd",
            "fontSize": 14,
          },
          Object {},
        ]
      }
      theme={
        Object {
          "fontColor": "#069ccd",
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "position": "absolute",
            "right": 0,
          },
        ]
      }
    />
  </View>
</View>
`;
